sim-lib/src/lib.rs

```diff
@@ -414,18 +453,58 @@ async fn produce_events(
 }
 
 async fn consume_simulation_results(
-    mut receiver: Receiver<(DispatchedPayment, PaymentResult)>,
-    shutdown: triggered::Trigger,
+    receiver: Receiver<(DispatchedPayment, PaymentResult)>,
+    _shutdown: triggered::Trigger,
+    listener: Listener,
 ) {
     log::debug!("Simulation results consumer started.");
 
-    while let Some(resolved_payment) = receiver.recv().await {
-        // TODO - write to CSV.
-        println!("Resolved payment received: {:?}", resolved_payment);
+    match log_payment_results(receiver, listener).await {
+        Ok(_) => {
+            log::debug!("Simulation results ok but exiting!!!");
+        }
+        Err(e) => log::error!("Error while reporting payment results: {:?}", e),
     }
+}
 
-    log::debug!("Simulation results consumer exiting");
-    shutdown.trigger();
+async fn log_payment_results(
+    mut receiver: Receiver<(DispatchedPayment, PaymentResult)>,
+    listener: Listener,
+) -> Result<(), SimulationError> {
+    let now = SystemTime::now()
+        .duration_since(SystemTime::UNIX_EPOCH)
+        .unwrap()
+        .as_secs();
+    let mut writer = WriterBuilder::new()
+        .buffer_capacity(size_of::<(DispatchedPayment, PaymentResult)>() * 5)
+        .from_path(format!("simulation_{:?}.csv", now))?;
+
+    loop {
+        tokio::select! {
+            biased;
+            _ = listener.clone() => {
+                log::debug!("Simulation results consumer received shutdown signal.");
+                break writer.flush().map_err(|_| SimulationError::FileError)
+            },
+            payment_report = receiver.recv() => {
+                match payment_report {
+                    Some(resolved_payment) => {
+                        log::info!("Resolved payment received: {:?}", resolved_payment.clone());
+                        writer.serialize(resolved_payment).map_err(|e| {
+                            let _ = writer.flush();
+                            SimulationError::CsvError(e)
+                        })?;
+                        log::debug!("Serialized payment result");
+                        continue;
+                    },
+                    None => {
+                        log::debug!("Received none outcome");
```

Comment:
I think we can remove this + the serialized debug logs once we're happy it's all werkin!